### PR TITLE
[Bugfix] Fix missing custom logit processor in SPEC_V2 verify

### LINKED ISSUE #21724: [Bug] custom_logit_processor is skipped in SPEC_V2 verify steps
### Checklist

- [ ] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [ ] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [ ] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [ ] Please use English. Otherwise, it will be closed.

### Describe the bug

When deploying `Qwen/Qwen3.5-35B-A3B-GPTQ-Int4` with speculative decoding enabled (`EAGLE`, with `SGLANG_ENABLE_SPEC_V2=1`), the custom logit processor appears to be applied only at the first generation step and is skipped on later verify steps.

This breaks reasoning-budget control in my deployment, because the processor is expected to run on every generation step, not just the first one.

After tracing the execution with debug logs, I found that:

- `apply_custom_logit_processor` is only executed once
- `TpModelWorker.forward_batch_generation(...)` is entered on every step, but on later steps it is called in verify mode and skips the downstream path that would normally reach sampling / logit processing

From the current code path, this does not look like only a logging artifact. It seems that the EAGLE v2 verify path does not execute the same custom-logit-processor logic that exists in the non-v2 path.

### What I observed

After adding debug logs, I saw that the custom logit processor is only applied once:

```text
[2026-03-30 20:25:43] Custom logit processor AdvancedReasoningLogitProcessor is applied.
```

I then traced the call path upward:

- `sglang.srt.model_executor.model_runner.ModelRunner.sample`
- `sglang.srt.managers.tp_workers.TpModelWorker.forward_batch_generation`

I added debug logs at the beginning of `forward_batch_generation(...)` and immediately before the `sample(...)` call.

The result was:

- `forward_batch_generation(...)` is entered on every step
- the branch that reaches `sample(...)` only happens once
- on the first step, `(model_worker_batch is not None) == True`
- on later steps, `(model_worker_batch is not None) == False`

Example debug check:

```python
class TpModelWorker:
    def forward_batch_generation(...):
        logger.debug(
            f"check, {self.pp_group.is_last_rank = }, {(model_worker_batch is not None) = }"
        )
```

From the function body, `model_worker_batch=None` is not necessarily invalid by itself, because the function does have logic to handle it. So this alone may not be the bug.

### Suspected root cause

I found two places in speculative verify code that explicitly pass `model_worker_batch=None` into `forward_batch_generation(...)`:

- `sglang.srt.speculative.eagle_worker_v2.EAGLEWorkerV2.verify`
- `sglang.srt.speculative.multi_layer_eagle_worker_v2.MultiLayerEagleWorkerV2.verify`

Relevant code:

- https://github.com/sgl-project/sglang/blob/0a1fb428697db7d52d9d1136fbeda5cb5adf781f/python/sglang/srt/speculative/eagle_worker_v2.py#L791-L798
- https://github.com/sgl-project/sglang/blob/0a1fb428697db7d52d9d1136fbeda5cb5adf781f/python/sglang/srt/speculative/multi_layer_eagle_worker_v2.py#L708-L715

The call looks like this:

```python
# Run target verify batch in the main compute stream (GPU compute)
forward_batch_output = self.target_worker.forward_batch_generation(
    model_worker_batch=None,
    forward_batch=verify_forward_batch,
    is_verify=True,
    skip_attn_bacckend_init=True,
)
```

I then traced the execution flow more carefully.

On the first step:

- `EAGLEWorkerV2.forward_batch_generation(...)` enters the `if` branch
- `TpModelWorker.forward_batch_generation(...)` reaches `self.model_runner.sample(...)`
- `ModelRunner.sample(...)` calls `self.sampler(...)`
- `Sampler.forward(...)`
- `Sampler._preprocess_logits(...)`
- `apply_custom_logit_processor(...)`

On later steps:

- `EAGLEWorkerV2.forward_batch_generation(...)` enters the `else` branch
- `EAGLEWorkerV2.verify(...)`
- `TpModelWorker.forward_batch_generation(...)` is called with `is_verify=True`
- because of the verify path, it skips the downstream logic that would normally reach `sample(...)`

I also checked where `apply_custom_logit_processor(...)` is used in speculative decoding:

- in `sglang.srt.speculative.eagle_worker.EAGLEWorker.verify`, `spec_info.verify(...)` is executed, which can reach the custom-logit-processor logic
- in `sglang.srt.speculative.eagle_worker_v2.EAGLEWorkerV2.verify`, I could not find an equivalent call path

In v2, the verify path seems to go through:

- `verify_input.prepare_for_v2_verify(...)`
- `verify_input.sample(...)`

These appear to come from `EagleVerifyInputV2Mixin`, and I could not find a path there that calls `EagleVerifyInput.verify(...)` or otherwise invokes `apply_custom_logit_processor(...)`.

So the current suspicion is:

- the problem is not simply that `model_worker_batch=None` is passed
- the more direct issue is that the EAGLE v2 verify path appears to bypass the custom-logit-processor logic entirely

### Expected behavior

- `custom_logit_processor` should run on verify steps as well when speculative decoding v2 is enabled
- the EAGLE v2 verify path should preserve the expected custom logit processor behavior
- reasoning-budget control should remain effective with EAGLE enabled

### Additional notes

- I am less certain now that this is a general `sample(...)` correctness problem; the stronger signal is specifically that the v2 verify path misses `apply_custom_logit_processor(...)`
- The `forward_batch_generation(...)` function signature may not fully reflect its real input contract, since the implementation appears to intentionally handle `model_worker_batch=None`
- Minor naming inconsistency noticed during tracing: `EAGLEWorkerV2` uses all-caps `EAGLE`, while `MultiLayerEagleWorkerV2` uses `Eagle`-style naming

### Reproduction

Deploy:

```bash
export SGLANG_USE_MODELSCOPE=true
export SGLANG_ENABLE_SPEC_V2=1
export SGLANG_MAMBA_CONV_DTYPE=float16

python -m sglang.launch_server \
  --model-path "Qwen/Qwen3.5-35B-A3B-GPTQ-Int4" \
  --served-model-name "qwen3.5-35b" \
  --dtype float16 \
  --mamba-scheduler-strategy extra_buffer \
  --enable-flashinfer-allreduce-fusion \
  --speculative-algorithm EAGLE \
  --speculative-eagle-topk 1 \
  --speculative-num-steps 3 \
  --speculative-num-draft-tokens 4 \
  --reasoning-parser qwen3 \
  --tool-call-parser qwen3_coder \
  --enable-custom-logit-processor \
  --enable-cache-report \
  --enable-metrics \
  --host 0.0.0.0 \
  --port 20080 \
  --log-level debug
```

Then send any request that uses `custom_logit_processor`.

### Environment

```
Python: 3.12.10 (main, Apr  9 2025, 04:03:51) [Clang 20.1.0 ]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA H20-3e
GPU 0,1,2,3,4,5,6,7 Compute Capability: 9.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.9, V12.9.86
CUDA Driver Version: 590.48.01
PyTorch: 2.9.1+cu128
sglang: 0.5.10rc0
sglang-kernel: 0.4.0
flashinfer_python: 0.6.6
flashinfer_cubin: 0.6.6
flashinfer_jit_cache: Module Not Found
triton: 3.5.1
transformers: 5.3.0
torchao: 0.9.0
numpy: 2.2.6
aiohttp: 3.13.4
fastapi: 0.135.2
huggingface_hub: 1.8.0
interegular: 0.3.3
modelscope: 1.35.3
orjson: 3.11.7
outlines: 0.1.11
packaging: 26.0
psutil: 7.2.2
pydantic: 2.13.0b2
python-multipart: 0.0.22
pyzmq: 27.1.0
uvicorn: 0.42.0
uvloop: 0.22.1
vllm: 0.16.0
xgrammar: 0.1.33
openai: 2.6.1
tiktoken: 0.12.0
anthropic: 0.86.0
litellm: Module Not Found
torchcodex: 0.9.1
NVIDIA Topology: 

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

ulimit soft: 1048576

```

### PR DESCRIPTION
**Pull request by aws-jacobyz**
## Motivation

Fixes #21724.

SPEC_V2 verify bypasses the normal `ModelRunner.sample(...)` path, so verify-time logits need to apply any required preprocessing directly. In the current implementation, `EagleVerifyInputV2Mixin.sample(...)` does not apply `custom_logit_processor`, which causes custom logit processors to be skipped during SPEC_V2 verify steps.

This restores SPEC_V2 verify behavior to match the existing SPEC_V1 verify path, where `EagleVerifyInput.verify(...)` already applies `custom_logit_processor(...)`.

## Modifications

- Apply `apply_custom_logit_processor(...)` in `python/sglang/srt/speculative/eagle_info_v2.py` during SPEC_V2 verify-time sampling.
- Add a focused unit regression test in `test/registered/unit/speculative/test_eagle_info_v2.py`.

## Accuracy Tests

- `python test/registered/unit/speculative/test_eagle_info_v2.py`

This is a small correctness fix with a focused unit regression test. I did not run a full model accuracy benchmark.

## Speed Tests and Profiling

- Not run
- No intended performance change

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

**Comment by jacobzhang22** [comment]
@hnyls2002 @Qiaolin-Yu @Ying1123 @merrymercy could you take a look at this PR when you get a chance? It's a small SPEC_V2 bugfix with a regression test. If it looks good, could someone also help trigger CI? Thanks!

**Comment by jacobzhang22** [comment]
I checked the remaining failures and they seem unrelated to the SPEC_V2 fix itself.

The relevant lint/stage-a checks are passing now. If helpful, could someone rerun or take a look at the affected jobs when convenient? Thanks!

**Comment by Qiaolin-Yu** [comment]
thanks for your contribution! I will review this later

**Comment by Qiaolin-Yu** [comment]
It seems the added test is failing (https://github.com/sgl-project/sglang/actions/runs/25344660965/job/74310629926?pr=21799#step:11:828), could you check this @jacobzhang22 ?

**Review by Qiaolin-Yu — COMMENTED** [review]
qq: do we also need this for spec v2? 
https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/speculative/eagle_info.py#L275

**Comment by aws-jacobyz** [comment]
Thanks for the review! I checked the failing test and pushed a follow-up fix. The stage-a CPU checks are green now.

For the `filter_batch(...)` part, I don’t think we need this for spec v2 because in spec v1 `bs` is taken from `self.retrieve_index.shape[0]`, so it handles the case where verify is operating on a kept subset and `sampling_info` needs to be filtered to match. In spec v2 `bs` is taken from `len(batch.seq_lens)`, and the verify input / retrieve metadata are rebuilt from the current batch each iteration.

**Comment by HH1162** [comment]
Hi @jacobzhang22 @Qiaolin-Yu,

I just hit the exact same issue and wanted to confirm your fix works. I independently reproduced the bug and verified this patch locally:

**Test environment:** Qwen3 (FP8) + NEXTN speculative decoding + `thinking_budget=1200`

| Before fix | After fix |
|---|---|
| reasoning_tokens: 3245 | reasoning_tokens: 1211 |
| budget not enforced | budget enforced (~1200) |

I actually submitted a duplicate PR (#26334) before realizing this one exists — already closed it.

This is a small but impactful fix. Would be great to see it merged. Thanks for the contribution!

**Comment by DreamGenX** [comment]
This bug still exists and this patch is worthwhile @Qiaolin-Yu @aws-jacobyz

**Comment by DreamGenX** [comment]
From my testing, this may result in duplicate `</think>` in output when speculative overlap scheduler is enabled, FWIW. @aws-jacobyz 

Related:
- https://github.com/sgl-project/sglang/issues/8056
- https://github.com/sgl-project/sglang/pull/20026
